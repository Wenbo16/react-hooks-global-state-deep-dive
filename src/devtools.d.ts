import { Enhancer } from './index';

export const reduxDevToolsExt: <S, A>() => Enhancer<S, A>;
