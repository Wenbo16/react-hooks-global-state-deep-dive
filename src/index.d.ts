export type StateItemUpdater<T> = (f: ((v: T) => T) | T) => void;

export type StateItemHook<T> = () => [T, StateItemUpdater<T>];

export const createGlobalState: <S extends {}>(initialState: S) => {
  stateItemUpdaters: { [K in keyof S]: StateItemUpdater<S[K]> },
  stateItemHooks: { [K in keyof S]: StateItemHook<S[K]> },
};
