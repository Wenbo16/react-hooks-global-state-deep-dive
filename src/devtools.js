/* eslint-env browser */

const compose = (...fns) => fns.reduce((p, c) => (...args) => p(c(...args)));

const initAction = () => ({ type: '@@redux/INIT' });

const createEnhancers = () => {
  let savedInitialState;
  const before = createStore => (reducer, initialState, enhancer) => {
    savedInitialState = initialState;
    if (enhancer) return enhancer(createStore)(reducer, initialState);
    const store = createStore(reducer, initialState);
    return {
      ...store,
      useGlobalState: (name) => {
        const [value] = store.useGlobalState(name);
        const MESG = 'Update is not allowed when using DevTools';
        return [value, () => { throw new Error(MESG); }];
      },
    };
  };
  const after = createStore => (reducer, initialState, enhancer) => {
    if (enhancer) return enhancer(createStore)(reducer, initialState);
    const store = createStore(null, savedInitialState);
    let devState = {
      ...reducer(initialState, initAction()),
      ...savedInitialState,
    };
    const getState = () => devState;
    const listeners = [];
    const keys = Object.keys(savedInitialState);
    const dispatch = (action) => {
      devState = reducer(devState, action);
      keys.forEach((key) => {
        const value = devState.computedStates[devState.currentStateIndex].state[key];
        const stateItem = store.stateItemMap[key];
        if (stateItem.getValue() !== value) {
          stateItem.updater(value);
        }
      });
      listeners.forEach(f => f());
      return action;
    };
    const subscribe = (listener) => {
      listeners.push(listener);
      const unsubscribe = () => {
        const index = listeners.indexOf(listener);
        listeners.splice(index, 1);
      };
      return unsubscribe;
    };
    return {
      ...store,
      getState,
      dispatch,
      subscribe,
    };
  };
  return { before, after };
};

export const reduxDevToolsExt = () => {
  if (!window.__REDUX_DEVTOOLS_EXTENSION__) return f => f;
  const { before, after } = createEnhancers();
  return compose(
    before,
    window.__REDUX_DEVTOOLS_EXTENSION__(),
    after,
  );
};
