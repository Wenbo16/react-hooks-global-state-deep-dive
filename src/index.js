import { useState, useEffect } from 'react';

const isFunction = fn => (typeof fn === 'function');

export const createGlobalState = (initialState) => {
  const globalState = { ...initialState };
  const stateItemListeners = {};
  const stateItemUpdaters = {};
  const stateItemHooks = {};
  Object.keys(globalState).forEach((name) => {
    stateItemListeners[name] = [];
    stateItemUpdaters[name] = (func) => {
      if (isFunction(func)) {
        globalState[name] = func(globalState[name]);
      } else {
        globalState[name] = func;
      }
      stateItemListeners[name].forEach(f => f(globalState[name]));
    };
    stateItemHooks[name] = () => {
      const [value, setValue] = useState(globalState[name]);
      useEffect(() => {
        stateItemListeners[name].push(setValue);
        const cleanup = () => {
          const index = stateItemListeners[name].indexOf(setValue);
          stateItemListeners[name].splice(index, 1);
        };
        return cleanup;
      }, []);
      return [value, stateItemUpdaters[name]];
    };
  });
  Object.freeze(stateItemUpdaters);
  Object.freeze(stateItemHooks);
  return { stateItemUpdaters, stateItemHooks };
};
