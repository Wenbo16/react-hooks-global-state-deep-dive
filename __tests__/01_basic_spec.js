/* eslint-env jest */

import React from 'react';
import { configure, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import Adapter from 'enzyme-adapter-react-16';

import { createGlobalState } from '../src/index';

configure({ adapter: new Adapter() });

describe('basic spec', () => {
  it('should have a function', () => {
    expect(createGlobalState).toBeDefined();
  });

  it.skip('should create a component with a global state', () => {
    const initialState = {
      counter1: 0,
    };
    const { stateItemHooks } = createGlobalState(initialState);
    const useCounter1 = stateItemHooks.counter1;
    const Counter = () => {
      const [value, update] = useCounter1();
      return (
        <div>
          <span>{value}</span>
          <button type="button" onClick={() => update(value + 1)}>+1</button>
        </div>
      );
    };
    const App = () => (
      <div>
        <div className="first">
          <Counter />
        </div>
        <div className="second">
          <Counter />
        </div>
      </div>
    );
    const wrapper = mount(<App />);
    expect(toJson(wrapper)).toMatchSnapshot();
    wrapper.find('.first button').simulate('click');
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
