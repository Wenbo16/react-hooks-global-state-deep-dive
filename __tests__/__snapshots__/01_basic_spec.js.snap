// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic spec should create a component with a global state 1`] = `
<App>
  <div>
    <div
      className="first"
    >
      <Counter>
        <div>
          <span>
            0
          </span>
          <button
            onClick={[Function]}
            type="button"
          >
            +1
          </button>
        </div>
      </Counter>
    </div>
    <div
      className="second"
    >
      <Counter>
        <div>
          <span>
            0
          </span>
          <button
            onClick={[Function]}
            type="button"
          >
            +1
          </button>
        </div>
      </Counter>
    </div>
  </div>
</App>
`;

exports[`basic spec should create a component with a global state 2`] = `
<App>
  <div>
    <div
      className="first"
    >
      <Counter>
        <div>
          <span>
            1
          </span>
          <button
            onClick={[Function]}
            type="button"
          >
            +1
          </button>
        </div>
      </Counter>
    </div>
    <div
      className="second"
    >
      <Counter>
        <div>
          <span>
            1
          </span>
          <button
            onClick={[Function]}
            type="button"
          >
            +1
          </button>
        </div>
      </Counter>
    </div>
  </div>
</App>
`;
